File: src/agy_swap/display.py
"""Output formatting, terminal display helpers, and duration utilities."""

from datetime import datetime, timezone, timedelta
import os
import re
import shutil
import sys
import unicodedata

from agy_swap import (
    ORANGE, GREEN, BLUE, RED, YELLOW, CYAN, GRAY, DARK_GRAY, BRIGHT_WHITE,
    BOLD, RESET, ANSI_ESCAPE, MAX_LIMIT_SECS, TIER_NAMES,
)


def configure_output():
    if sys.stdout.isatty() and "NO_COLOR" not in os.environ:
        return
    for name in ("ORANGE", "GREEN", "BLUE", "RED", "YELLOW", "CYAN", "GRAY", "DARK_GRAY", "BRIGHT_WHITE", "BOLD", "RESET"):
        globals()[name] = ""


def clean_display_text(value, fallback=""):
    text = str(value or fallback)
    return "".join(ch for ch in text if ch in "\t " or unicodedata.category(ch)[0] != "C").strip()


def normalize_email(value):
    email = clean_display_text(value).lower()
    if not re.fullmatch(r"[^@\s]+@[^@\s]+\.[^@\s]+", email):
        return None
    return email


def _parse_utc_datetime(value):
    dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
    if dt.tzinfo is None:
        dt = dt.replace(tzinfo=timezone.utc)
    return dt.astimezone(timezone.utc)


def format_duration(seconds):
    mins = max(0, int(seconds // 60))
    days, mins = divmod(mins, 24 * 60)
    hours, mins = divmod(mins, 60)
    if days:
        return f"{days}d {hours}h {mins}m"
    return f"{hours}h {mins}m" if hours else f"{mins}m"


def parse_duration_seconds(duration_str):
    if not duration_str:
        return None
    duration_str = duration_str.strip().lower()
    if duration_str in ("reset", "clear", "0", "none"):
        return 0
    if duration_str.isdigit():
        total = int(duration_str) * 60
    else:
        match = re.fullmatch(r"\s*(?:(\d+)\s*d)?\s*(?:(\d+)\s*h)?\s*(?:(\d+)\s*m)?\s*(?:(\d+)\s*s)?\s*", duration_str)
        if not match or not any(match.groups()):
            return None
        days, hours, mins, secs = (int(value or 0) for value in match.groups())
        total = days * 86400 + hours * 3600 + mins * 60 + secs
    return total if 0 <= total <= MAX_LIMIT_SECS else None


def quota_groups(acc):
    snapshot = acc.get("quota_snapshot")
    groups = snapshot.get("groups") if isinstance(snapshot, dict) else None
    return groups if isinstance(groups, list) else []


def format_quota_bar(bucket, now=None, width=12):
    fraction = min(1.0, max(0.0, float(bucket["remaining_fraction"])))
    filled = round(fraction * width)
    color = RED if fraction <= 0.1 else YELLOW if fraction <= 0.3 else GREEN
    bar = f"{color}{'█' * filled}{DARK_GRAY}{'░' * (width - filled)}{RESET}"
    try:
        remaining = (_parse_utc_datetime(bucket["reset_at"]) - (now or datetime.now(timezone.utc))).total_seconds()
        reset = f" · resets in {format_duration(remaining)}" if remaining > 0 else " · refresh due"
    except (KeyError, AttributeError, TypeError, ValueError):
        reset = ""
    return f"[{bar}] {fraction * 100:.2f}% remaining{reset}"


def quota_wait_seconds(acc, now=None, family=None):
    groups = quota_groups(acc)
    if not groups:
        return None
    now = now or datetime.now(timezone.utc)
    group_id = "gemini" if family == "gemini" else "third_party" if family in ("claude", "gpt") else None
    matched = False
    waits = []
    for group in groups:
        if group_id and group.get("id") != group_id:
            continue
        matched = True
        for bucket in group.get("buckets", []):
            if bucket.get("remaining_fraction", 1) > 0:
                continue
            try:
                waits.append(max(0, (_parse_utc_datetime(bucket["reset_at"]) - now).total_seconds()))
            except (KeyError, AttributeError, TypeError, ValueError):
                pass
    return max(waits, default=0) if matched else None


def format_cooldown_bar(limit, now=None, width=12):
    if not limit:
        return f"[{DARK_GRAY}{'░' * width}{RESET}] No active cooldown observed"
    now = now or datetime.now(timezone.utc)
    try:
        observed_at = _parse_utc_datetime(limit["observed_at"])
        reset_at = _parse_utc_datetime(limit["reset_at"])
    except (KeyError, TypeError, ValueError):
        return None
    total = (reset_at - observed_at).total_seconds()
    if total <= 0:
        return None
    ratio = min(1.0, max(0.0, (reset_at - now).total_seconds() / total))
    filled = round(ratio * width)
    bar = f"{RED}{'█' * filled}{DARK_GRAY}{'░' * (width - filled)}{RESET}"
    return f"[{bar}] {ratio * 100:.1f}% time left"


def active_quota_limits(acc, now=None):
    now = now or datetime.now(timezone.utc)
    active = []
    for limit in acc.get("quota_limits", {}).values():
        try:
            reset_at = _parse_utc_datetime(limit["reset_at"])
        except (KeyError, TypeError, ValueError):
            continue
        remaining = (reset_at - now).total_seconds()
        if remaining > 0:
            active.append((remaining, limit))
    return sorted(active, key=lambda item: item[0])


def get_tier_badge(acc):
    snapshot = acc.get("quota_snapshot")
    tier = snapshot.get("tier") if isinstance(snapshot, dict) else None
    if isinstance(tier, dict) and tier.get("name"):
        plan = clean_display_text(tier["name"])
        color = GRAY if tier.get("id") == "free-tier" else ORANGE
        return f"{color}{plan}{RESET}"
    source = acc.get("tier_source")
    plan = clean_display_text(acc.get("plan") or acc.get("tier")).title()
    if plan not in ("Pro", "Starter", "Free") or source != "manual":
        return f"{GRAY}Unknown{RESET}"
    color = ORANGE if plan == "Pro" else GRAY
    return f"{color}{plan} (manual){RESET}"


def get_account_limit_display(acc, is_active=False, current_token_str=None):
    tier = get_tier_badge(acc)
    buckets = [bucket for group in quota_groups(acc) for bucket in group.get("buckets", [])]
    if buckets:
        lowest = min(buckets, key=lambda bucket: bucket["remaining_fraction"])
        fraction = lowest["remaining_fraction"]
        if fraction <= 0:
            try:
                remaining = (_parse_utc_datetime(lowest["reset_at"]) - datetime.now(timezone.utc)).total_seconds()
            except (KeyError, AttributeError, TypeError, ValueError):
                remaining = 0
            return f"[{tier}] {RED}Limited{RESET} · {lowest['name']} resets in {format_duration(remaining)}"
        color = RED if fraction <= 0.1 else YELLOW if fraction <= 0.3 else GREEN
        return f"[{tier}] {color}{fraction * 100:.0f}% lowest remaining{RESET}"
    active = active_quota_limits(acc)
    if active:
        parts = [f"{RED}! {limit['model']} ({format_duration(remaining)}){RESET}" for remaining, limit in active]
        return f"[{tier}] " + " ".join(parts)
    return f"[{tier}] {GRAY}Usage unavailable · no recent cooldown error{RESET}"


def get_avatar_badge(name, email):
    name_str = clean_display_text(name)
    parts = name_str.split()
    if len(parts) >= 2 and parts[0] and parts[1]:
        initials = (parts[0][0] + parts[1][0]).upper()
    elif len(parts) == 1 and parts[0]:
        initials = parts[0][:2].upper()
    else:
        initials = email[:2].upper() if email else "GU"

    if not sys.stdout.isatty() or "NO_COLOR" in os.environ:
        return initials

    bg_colors = [166, 172, 64, 71, 133, 167, 31, 68]
    bg_code = bg_colors[sum(email.encode("utf-8")) % len(bg_colors)]
    return f"\033[48;5;{bg_code}m\033[38;5;255m\033[1m {initials} \033[0m"


def visible_len(text):
    return sum(
        0 if unicodedata.combining(ch) else 2 if unicodedata.east_asian_width(ch) in ("W", "F") else 1
        for ch in ANSI_ESCAPE.sub("", text)
    )


def truncate_visible(text, width):
    if visible_len(text) <= width:
        return text
    result = []
    used = 0
    index = 0
    while index < len(text) and used < max(0, width - 1):
        match = ANSI_ESCAPE.match(text, index)
        if match:
            result.append(match.group(0))
            index = match.end()
            continue
        ch = text[index]
        char_width = 0 if unicodedata.combining(ch) else 2 if unicodedata.east_asian_width(ch) in ("W", "F") else 1
        if used + char_width > width - 1:
            break
        result.append(ch)
        used += char_width
        index += 1
    return "".join(result) + "…" + RESET


def get_responsive_width():
    try:
        cols, _ = shutil.get_terminal_size((80, 24))
        return max(28, min(cols - 2, 110))
    except Exception:
        return 76

File: internal/app/display.go
package app

import (
	"fmt"
	"math"
	"os"
	"regexp"
	"sort"
	"strings"
	"time"
	"unicode"
	"unicode/utf8"
)

type palette struct {
	Orange, Green, Blue, Red, Yellow, Cyan, Gray, DarkGray, White, Bold, Reset string
}

func makePalette(color bool) palette {
	if !color || os.Getenv("NO_COLOR") != "" {
		return palette{}
	}
	return palette{Orange: "\x1b[38;5;208m", Green: "\x1b[38;5;78m", Blue: "\x1b[38;5;75m", Red: "\x1b[38;5;203m", Yellow: "\x1b[38;5;220m", Cyan: "\x1b[38;5;86m", Gray: "\x1b[38;5;244m", DarkGray: "\x1b[38;5;238m", White: "\x1b[38;5;255m", Bold: "\x1b[1m", Reset: "\x1b[0m"}
}

func formatDuration(seconds float64) string {
	minutes := int(math.Floor(seconds / 60))
	if minutes < 0 {
		minutes = 0
	}
	days := minutes / (24 * 60)
	minutes %= 24 * 60
	hours := minutes / 60
	minutes %= 60
	if days > 0 {
		return fmt.Sprintf("%dd %dh %dm", days, hours, minutes)
	}
	if hours > 0 {
		return fmt.Sprintf("%dh %dm", hours, minutes)
	}
	return fmt.Sprintf("%dm", minutes)
}

var durationPattern = regexp.MustCompile(`^\s*(?:(\d+)\s*d)?\s*(?:(\d+)\s*h)?\s*(?:(\d+)\s*m)?\s*(?:(\d+)\s*s)?\s*$`)

func parseDuration(value string) (time.Duration, bool) {
	value = strings.ToLower(strings.TrimSpace(value))
	if oneOf(value, "reset", "clear", "0", "none") {
		return 0, true
	}
	if value == "" {
		return 0, false
	}
	if allDigits(value) {
		var mins int64
		if _, err := fmt.Sscan(value, &mins); err != nil {
			return 0, false
		}
		d := time.Duration(mins) * time.Minute
		return d, d <= maxLimitDuration
	}
	match := durationPattern.FindStringSubmatch(value)
	if match == nil || strings.Join(match[1:], "") == "" {
		return 0, false
	}
	var values [4]int64
	for i := range values {
		if match[i+1] != "" {
			_, _ = fmt.Sscan(match[i+1], &values[i])
		}
	}
	d := time.Duration(values[0])*24*time.Hour + time.Duration(values[1])*time.Hour + time.Duration(values[2])*time.Minute + time.Duration(values[3])*time.Second
	return d, d >= 0 && d <= maxLimitDuration
}

func allDigits(s string) bool {
	for _, r := range s {
		if r < '0' || r > '9' {
			return false
		}
	}
	return s != ""
}

type ActiveLimit struct {
	Remaining time.Duration
	Limit     map[string]any
}

func activeLimits(account Account, now time.Time) []ActiveLimit {
	var result []ActiveLimit
	for _, raw := range getMap(account["quota_limits"]) {
		limit := getMap(raw)
		reset, err := parseUTC(getString(limit, "reset_at"))
		if err != nil {
			continue
		}
		remaining := reset.Sub(now)
		if remaining > 0 {
			result = append(result, ActiveLimit{remaining, limit})
		}
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Remaining < result[j].Remaining })
	return result
}

func quotaGroups(account Account) []any {
	snapshot := getMap(account["quota_snapshot"])
	return getSlice(snapshot["groups"])
}

func quotaWait(account Account, now time.Time, family string) (time.Duration, bool) {
	groups := quotaGroups(account)
	if len(groups) == 0 {
		return 0, false
	}
	groupID := ""
	if family == "gemini" {
		groupID = "gemini"
	} else if family == "claude" || family == "gpt" {
		groupID = "third_party"
	}
	matched := false
	var maxWait time.Duration
	for _, raw := range groups {
		group := getMap(raw)
		if groupID != "" && getString(group, "id") != groupID {
			continue
		}
		matched = true
		for _, rb := range getSlice(group["buckets"]) {
			b := getMap(rb)
			f, _ := getFloat(b["remaining_fraction"])
			if f > 0 {
				continue
			}
			reset, e := parseUTC(getString(b, "reset_at"))
			if e == nil && reset.Sub(now) > maxWait {
				maxWait = max(0, reset.Sub(now))
			}
		}
	}
	return maxWait, matched
}

func tierBadge(account Account, p palette) string {
	snapshot := getMap(account["quota_snapshot"])
	tier := getMap(snapshot["tier"])
	if name := cleanText(getString(tier, "name")); name != "" {
		c := p.Orange
		if getString(tier, "id") == "free-tier" {
			c = p.Gray
		}
		return c + name + p.Reset
	}
	plan := strings.Title(strings.ToLower(cleanText(firstString(account["plan"], getString(account, "tier")))))
	if getString(account, "tier_source") != "manual" || !oneOf(plan, "Pro", "Starter", "Free") {
		return p.Gray + "Unknown" + p.Reset
	}
	c := p.Gray
	if plan == "Pro" {
		c = p.Orange
	}
	return c + plan + " (manual)" + p.Reset
}

type quotaGroupHealth struct {
	id       string
	label    string
	bucket   string
	fraction float64
	resetAt  time.Time
	found    bool
}

func quotaGroupHealths(account Account) []quotaGroupHealth {
	groups := quotaGroups(account)
	result := make([]quotaGroupHealth, 0, len(groups))
	for _, rawGroup := range groups {
		group := getMap(rawGroup)
		health := quotaGroupHealth{
			id:    strings.ToLower(cleanText(getString(group, "id"))),
			label: quotaGroupLabel(group),
		}
		for _, rawBucket := range getSlice(group["buckets"]) {
			bucket := getMap(rawBucket)
			fraction, ok := getFloat(bucket["remaining_fraction"])
			if !ok {
				continue
			}
			fraction = max(0, min(1, fraction))
			if !health.found || fraction < health.fraction {
				health.found = true
				health.fraction = fraction
				health.bucket = cleanText(getString(bucket, "name"))
				health.resetAt, _ = parseUTC(getString(bucket, "reset_at"))
			}
		}
		if health.found {
			result = append(result, health)
		}
	}
	return result
}

func quotaGroupLabel(group map[string]any) string {
	switch strings.ToLower(cleanText(getString(group, "id"))) {
	case "gemini":
		return "Gemini"
	case "third_party":
		return "Claude/GPT"
	}
	name := cleanText(getString(group, "name"))
	name = strings.TrimSpace(strings.TrimSuffix(strings.TrimSuffix(name, " Models"), " models"))
	return firstString(name, "Quota")
}

func bestAvailableQuotaGroup(groups []quotaGroupHealth) (quotaGroupHealth, bool) {
	var best quotaGroupHealth
	found := false
	for _, group := range groups {
		if group.fraction <= 0 {
			continue
		}
		if !found || group.fraction > best.fraction {
			best, found = group, true
		}
	}
	return best, found
}

func limitedQuotaGroupLabels(groups []quotaGroupHealth) []string {
	labels := make([]string, 0, len(groups))
	for _, group := range groups {
		if group.fraction <= 0 {
			labels = append(labels, group.label)
		}
	}
	return labels
}

func accountStatus(account Account, p palette, now time.Time) string {
	tier := tierBadge(account, p)
	groups := quotaGroupHealths(account)
	if best, ok := bestAvailableQuotaGroup(groups); ok {
		color := p.Green
		if best.fraction <= .1 {
			color = p.Red
		} else if best.fraction <= .3 {
			color = p.Yellow
		}
		status := fmt.Sprintf("[%s] %sReady%s · %s %.2f%% available", tier, color, p.Reset, best.label, best.fraction*100)
		if limited := limitedQuotaGroupLabels(groups); len(limited) > 0 {
			status += " · " + strings.Join(limited, "/") + " limited"
		}
		return status
	}
	if len(groups) > 0 {
		limited := groups[0]
		for _, group := range groups[1:] {
			if !limited.resetAt.IsZero() && !group.resetAt.IsZero() && group.resetAt.Before(limited.resetAt) {
				limited = group
			}
		}
		reset := ""
		if !limited.resetAt.IsZero() && limited.resetAt.After(now) {
			reset = fmt.Sprintf(" · %s resets in %s", limited.bucket, formatDuration(limited.resetAt.Sub(now).Seconds()))
		}
		return fmt.Sprintf("[%s] %sLimited%s%s", tier, p.Red, p.Reset, reset)
	}
	limits := activeLimits(account, now)
	if len(limits) > 0 {
		parts := make([]string, 0, len(limits))
		for _, l := range limits {
			parts = append(parts, fmt.Sprintf("%s! %s (%s)%s", p.Red, getString(l.Limit, "model"), formatDuration(l.Remaining.Seconds()), p.Reset))
		}
		return "[" + tier + "] " + strings.Join(parts, " ")
	}
	return fmt.Sprintf("[%s] %sUsage unavailable · no recent cooldown error%s", tier, p.Gray, p.Reset)
}

func formatQuotaBar(bucket map[string]any, p palette, now time.Time, width int) string {
	fraction, _ := getFloat(bucket["remaining_fraction"])
	fraction = max(0, min(1, fraction))
	filled := int(math.Round(fraction * float64(width)))
	color := p.Green
	if fraction <= .1 {
		color = p.Red
	} else if fraction <= .3 {
		color = p.Yellow
	}
	bar := color + strings.Repeat("█", filled) + p.DarkGray + strings.Repeat("░", width-filled) + p.Reset
	reset := ""
	if at, err := parseUTC(getString(bucket, "reset_at")); err == nil {
		remaining := at.Sub(now)
		if remaining > 0 {
			reset = " · resets in " + formatDuration(remaining.Seconds())
		} else {
			reset = " · refresh due"
		}
	}
	return fmt.Sprintf("[%s] %.2f%% remaining%s", bar, fraction*100, reset)
}

// formatQuotaBarResponsive keeps the value and reset window visible when a
// stacked/compact detail pane cannot fit the full desktop copy. The bar is
// still rendered with the same palette and fraction; only the copy gets
// progressively shorter.
func formatQuotaBarResponsive(bucket map[string]any, p palette, now time.Time, width, maxWidth int) string {
	width = maxInt(1, width)
	maxWidth = maxInt(1, maxWidth)
	fraction, _ := getFloat(bucket["remaining_fraction"])
	fraction = max(0, min(1, fraction))
	filled := int(math.Round(fraction * float64(width)))
	color := p.Green
	if fraction <= .1 {
		color = p.Red
	} else if fraction <= .3 {
		color = p.Yellow
	}
	bar := color + strings.Repeat("█", filled) + p.DarkGray + strings.Repeat("░", width-filled) + p.Reset
	remaining := ""
	if at, err := parseUTC(getString(bucket, "reset_at")); err == nil {
		if left := at.Sub(now); left > 0 {
			remaining = formatDuration(left.Seconds())
		} else {
			remaining = "due"
		}
	}
	full := fmt.Sprintf("[%s] %.2f%% remaining", bar, fraction*100)
	if remaining != "" {
		full += " · resets in " + remaining
	}
	if visibleWidth(full) <= maxWidth {
		return full
	}
	if remaining != "" && maxWidth >= width+16 {
		short := remaining
		if fields := strings.Fields(short); len(fields) > 2 {
			short = strings.Join(fields[:2], " ")
		}
		candidate := fmt.Sprintf("[%s] %.1f%% · %s", bar, fraction*100, short)
		if visibleWidth(candidate) <= maxWidth {
			return candidate
		}
	}
	return fmt.Sprintf("[%s] %.0f%%", bar, fraction*100)
}

func formatCooldownBar(limit map[string]any, p palette, now time.Time, width int) string {
	observed, e1 := parseUTC(getString(limit, "observed_at"))
	reset, e2 := parseUTC(getString(limit, "reset_at"))
	if e1 != nil || e2 != nil || !reset.After(observed) {
		return ""
	}
	ratio := max(0, min(1, reset.Sub(now).Seconds()/reset.Sub(observed).Seconds()))
	filled := int(math.Round(ratio * float64(width)))
	bar := p.Red + strings.Repeat("█", filled) + p.DarkGray + strings.Repeat("░", width-filled) + p.Reset
	return fmt.Sprintf("[%s] %.1f%% time left", bar, ratio*100)
}

func avatar(name, email string, color bool) string {
	parts := strings.Fields(cleanText(name))
	initials := "GU"
	if len(parts) >= 2 {
		initials = strings.ToUpper(firstRune(parts[0]) + firstRune(parts[1]))
	} else if len(parts) == 1 {
		r := []rune(parts[0])
		if len(r) > 2 {
			r = r[:2]
		}
		initials = strings.ToUpper(string(r))
	} else if len(email) >= 2 {
		initials = strings.ToUpper(email[:2])
	}
	label := "[" + initials + "]"
	if !color || os.Getenv("NO_COLOR") != "" {
		return label
	}
	colors := []int{166, 172, 64, 71, 133, 167, 31, 68}
	sum := 0
	for _, b := range []byte(email) {
		sum += int(b)
	}
	// Keep the avatar ASCII-shaped. Background-color blocks are attractive in
	// some terminals but are captured as replacement glyphs or bleed into the
	// next column in others, which makes the TUI table look corrupted.
	return fmt.Sprintf("\x1b[38;5;%dm\x1b[1m%s\x1b[0m", colors[sum%len(colors)], label)
}

func firstRune(s string) string {
	for _, r := range s {
		return string(r)
	}
	return ""
}

var ansiPattern = regexp.MustCompile(`\x1B(?:[@-Z\\-_]|\[[0-?]*[ -/]*[@-~])`)

func visibleWidth(s string) int {
	n := 0
	for _, r := range ansiPattern.ReplaceAllString(s, "") {
		if unicode.Is(unicode.Mn, r) {
			continue
		}
		if isWide(r) {
			n += 2
		} else {
			n++
		}
	}
	return n
}
func isWide(r rune) bool {
	return r >= 0x1100 && (r <= 0x115f || r == 0x2329 || r == 0x232a || r >= 0x2e80 && r <= 0xa4cf || r >= 0xac00 && r <= 0xd7a3 || r >= 0xf900 && r <= 0xfaff || r >= 0xfe10 && r <= 0xfe19 || r >= 0xfe30 && r <= 0xfe6f || r >= 0xff00 && r <= 0xff60 || r >= 0xffe0 && r <= 0xffe6 || r >= 0x1f300)
}
func truncateVisible(s string, width int, p palette) string {
	if width <= 0 {
		return ""
	}
	if visibleWidth(s) <= width {
		return s
	}
	var b strings.Builder
	used := 0
	limit := maxInt(0, width-1)
	for len(s) > 0 {
		if match := ansiPattern.FindStringIndex(s); match != nil && match[0] == 0 {
			b.WriteString(s[:match[1]])
			s = s[match[1]:]
			continue
		}
		r, size := utf8.DecodeRuneInString(s)
		if size == 0 {
			break
		}
		w := 1
		if unicode.Is(unicode.Mn, r) {
			w = 0
		} else if isWide(r) {
			w = 2
		}
		if used+w > limit {
			break
		}
		b.WriteString(s[:size])
		used += w
		s = s[size:]
	}
	return b.String() + "…" + p.Reset
}
